File: SortVision/public/code/radix/python/radixSort.py
"""Radix Sort (LSD, base 10, non-negative integers).

Time: O(d * (n + k))
Space: O(n + k)
"""

def radix_sort(arr: list[int]) -> list[int]:
    if len(arr) <= 1:
        return arr

    max_val = max(arr)
    exp = 1
    while max_val // exp > 0:
        _counting_sort_by_exp(arr, exp)
        exp *= 10

    return arr


def _counting_sort_by_exp(arr: list[int], exp: int) -> None:
    output = [0] * len(arr)
    count = [0] * 10

    for x in arr:
        count[(x // exp) % 10] += 1
    for i in range(1, 10):
        count[i] += count[i - 1]

    for i in range(len(arr) - 1, -1, -1):
        digit = (arr[i] // exp) % 10
        output[count[digit] - 1] = arr[i]
        count[digit] -= 1

    arr[:] = output


if __name__ == "__main__":
    arr = [170, 45, 75, 90, 802, 24, 2, 66]
    print(radix_sort(arr))

File: SortVision/public/code/radix/golang/radixSort.go
package main

import "fmt"

// Radix Sort (LSD, base 10, non-negative integers)
// Time: O(d * (n + k)), Space: O(n + k)
func radixSort(arr []int) {
	if len(arr) <= 1 {
		return
	}

	maxVal := arr[0]
	for _, x := range arr {
		if x > maxVal {
			maxVal = x
		}
	}

	for exp := 1; maxVal/exp > 0; exp *= 10 {
		countingSortByExp(arr, exp)
	}
}

func countingSortByExp(arr []int, exp int) {
	n := len(arr)
	output := make([]int, n)
	count := make([]int, 10)

	for _, x := range arr {
		count[(x/exp)%10]++
	}
	for i := 1; i < 10; i++ {
		count[i] += count[i-1]
	}

	for i := n - 1; i >= 0; i-- {
		digit := (arr[i] / exp) % 10
		output[count[digit]-1] = arr[i]
		count[digit]--
	}

	copy(arr, output)
}

func main() {
	arr := []int{170, 45, 75, 90, 802, 24, 2, 66}
	radixSort(arr)
	fmt.Println(arr)
}
